Fix Dataset URI by using openlineage-integration-common as lib

tatiana opened the pull request:
Use `openlineage-dbt` to create outlets Dataset URIs from within Cosmos.

Closes: #305
Closes: #497
Closes: #433 (only emits outlet events from the model - the same behaviour as openlineage-dbt)

Validation
This change was tested by running Marquez locally and triggering the dag `basic_cosmos_dag` using Airflow 2.6.1 and Python 3.10.10.

The output generated by this version of Cosmos can be seen in the following screenshots:

<img width="1624" alt="Screenshot 2023-09-04 at 22 08 32" src="https://github.com/astronomer/astronomer-cosmos/assets/272048/bf8ac0fe-a4de-42a1-aac3-4e111876b615">

<img width="1624" alt="Screenshot 2023-09-04 at 22 09 01" src="https://github.com/astronomer/astronomer-cosmos/assets/272048/3d41821b-2cfb-414d-9951-e0b062ba0a1a">

<img width="1624" alt="Screenshot 2023-09-04 at 22 09 31" src="https://github.com/astronomer/astronomer-cosmos/assets/272048/1ad93331-70b1-428d-b638-e73d26492c96">


Tasks
- [x] Fix pre-commit checks
- [x] Add test
- [x]  Inlets support
- [x]  ~Emit open lineage events~
- [x]  ~Support Docker/K8s~ (deverred to issue: #496)
- [x]  ~Create a PR on openlineage-dbt to remove the dbt dependency~ Not needed since the code depends on `openlineage-integration-common` and not `openlineage-dbt` ([more info](https://github.com/OpenLineage/OpenLineage/blob/81372ca2bc2afecab369eab4a54cc6380dda49d0/integration/common/setup.py#L15))
- [x]  Understand which dataset is being emitted from test tasks (only inlets, no outlets)

tatiana commented on tests/operators/test_local.py:
```diff
@@ -124,6 +139,68 @@ def test_dbt_base_operator_get_env(p_context_to_airflow_vars: MagicMock) -> None
     assert env == expected_env
 
 
+# TATI IS HERE
```
Remove this comment 🙈

mobuchowski commented on cosmos/operators/local.py:
```diff
@@ -132,9 +155,10 @@ def run_subprocess(self, *args: Any, **kwargs: Any) -> FullOutputSubprocessResul
+        cmd: list[str],
         env: dict[str, str | bytes | os.PathLike[Any]],
         context: Context,
+        emit_openlineage: bool = False,
```
Why `emit_openlineage`? 

If OpenLineage integration exists and is properly configured, it will call `get_openlineage_facets_on_complete` itself. I don't think there's need to have additional explicit boolean check for it, although I don't fully know the context.

tatiana replied:
By default, it seems the common processor would raise an exception if it is an unsupported dbt node:
https://github.com/OpenLineage/OpenLineage/blob/main/integration/common/openlineage/common/provider/dbt/processor.py#L141

I'll change our code to set:
```
    should_raise_on_unsupported_command = False
```

JDarDagran commented on cosmos/operators/local.py:
```diff
@@ -5,14 +5,34 @@
+    from openlineage.common.provider.dbt.local import DbtLocalArtifactProcessor
+    from airflow.datasets import Dataset
+except ModuleNotFoundError:
+    is_open_lineage_available = False
```
small thing but `is_openlineage_available` to be consistent with naming

harels commented on cosmos/operators/local.py:
```diff
@@ -189,13 +220,97 @@ def run_command(
+            os.environ[key] = str(value)
+
+        openlineage_processor = DbtLocalArtifactProcessor(
+            producer="https://github.com/astronomer/astronomer-cosmos/",
```
should this come from project consts?

harels reviewed (approved):
looks good, thanks @tatiana.
Left some comments.
2 questions: 
1. do we have tests that verify the output of the openlineage events themselves? 
2. do we need to update docs as well?

tatiana commented:
Thanks a lot for the review and feedback, @harels @mobuchowski and @JDarDagran ! I've addressed most of the feedback. As @harels mentioned, one point that is missing is to have a better automated testing strategy to make sure Cosmos is emitting the open lineage events correctly. I've logged a follow up ticket for us to improve this: #509 509
